nvtabular/dispatch.py

```diff
@@ -168,3 +190,76 @@ def _to_arrow(x):
         return x.to_arrow()
     else:
         return pa.Table.from_pandas(x, preserve_index=False)
+
+
+def _detect_format(data):
+    """Utility to detect the format of `data`"""
+    from nvtabular import Dataset
+
+    if isinstance(data, Dataset):
+        return ExtData.DATASET
+    elif isinstance(data, dd.DataFrame):
+        if isinstance(data._meta, cudf.DataFrame):
+            return ExtData.DASK_CUDF
+        return ExtData.DASK_PANDAS
+    elif isinstance(data, cudf.DataFrame):
+        return ExtData.CUDF
+    elif isinstance(data, pd.DataFrame):
+        return ExtData.PANDAS
+    elif isinstance(data, pa.Table):
+        return ExtData.ARROW
+    else:
+        mapping = {
+            "pq": ExtData.PARQUET,
+            "parquet": ExtData.PARQUET,
+            "csv": ExtData.CSV,
+        }
+        if isinstance(data, list) and data:
+            file_type = mapping.get(str(data[0]).split(".")[-1], None)
+        else:
+            file_type = mapping.get(str(data).split(".")[-1], None)
+        if file_type is None:
+            raise ValueError("Data format not recognized.")
+        return file_type
+
+
+def _convert_data(x, cpu=True):
+    """Move data between cpu and gpu-backed data.
+
+    Note that the input ``x`` may be an Arrow Table,
+    but the output will only be a pandas or cudf DataFrame.
+    """
+    if cpu:
+        if isinstance(x, dd.DataFrame):
+            # If input is a dask_cudf collection, convert
+            # to a pandas-based Dask collection
+            if not isinstance(x, dask_cudf.DataFrame):
+                return x
+            return x.to_dask_dataframe()
+        else:
+            if isinstance(x, pd.DataFrame):
+                return x
+            return x.to_pandas()
+    else:
+        if isinstance(x, dd.DataFrame):
+            # If input is a Dask collection, covert to dask_cudf
+            if isinstance(x, dask_cudf.DataFrame):
+                return x
+            return x.map_partitions(cudf.from_pandas)
+        elif isinstance(x, pa.Table):
+            return cudf.DataFrame.from_arrow(x)
+        else:
+            if isinstance(x, cudf.DataFrame):
+                return x
+            return cudf.DataFrame.from_pandas(x)
```

Comment:
Would it be possible to use the existing nvtabular dataset code here?  Like we have code to convert from parquet/csv/dask/pandas/cudf/ here https://github.com/NVIDIA/NVTabular/blob/94d2960c95a6b9bef71483c80d80bf212429af32/nvtabular/io/dataset.py#L228-L269 . I'm wondering if we can just use the Dataset to convert to a known format in the JoinExternal op, and avoid having to have an extra conversion step here

Reply:
Good point - The logic should probably be shared, and live in dispatch.py for both of these cases (so we can eventually isolate explicit references to cudf/dask_cudf).

Reply:
I moved the shared Dataset logic into `dispatch.py` - let me know if you think this makes sense.  I am hoping to eventually consolidate all cudf-specific logic into one location (like dispatch).

Reply:
nice! thanks